### DISCUSSION TITLE
constant index model

### DISCUSSION THREAD

**Discussion by TLawu** [mjhoptics/ray-optics#92] [Q&A] [answered]
I am new with ray-optics (python as well).
I tried to create a simple optical model, and use examples for references.
I want to add surface with constant index.
What I did to assign radius 7.77, thickness 0.55, refractive_index 1.493, semi-diameter 5.0 was

sm.add_surface([7.77, 0.55, 1.493, '', 5.0])

But this will give error
TypeError: unsupported operand type(s) for /: 'float' and 'str'
The problem seems to be the v-number type.

If I put v-number value, it goes well.

How to add surface with constant refractive_index?
Should I put very big value for v-number?

Thanks in advance for any help.

**Comment by TLawu** [comment]
Just to make it clear, what I want to do is to make a non-wavelength dependency analyses.
The optical model I want to build is a simple eye model as illustrated in the following figure.
![model](https://user-images.githubusercontent.com/109012617/180633574-72d80df9-ed07-4d45-9163-0377d6c66ec9.png)

My code
```
from rayoptics.environment import *
opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']

osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=3)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=5.0, flds=[0.],is_relative=True)
osp['wvls'] = WvlSpec([(546.074, 1.0)], ref_wl=0)

opm.radius_mode = True
sm.gaps[0].thi=1e10
sm.add_surface([7.77, 0.55, 1.376, '', 5.])
sm.add_surface([6.8, 3.4, 1.336, '', 5.])
sm.add_surface([0., 1.6, 1.336, '', 3.])
sm.set_stop()
sm.add_surface([20., 0.7, 1.5, '', 3.])
sm.add_surface([-20., 21., 1.336, '', 3.])
sm.add_surface([0., 0.])

sm.do_apertures = False
opm.update_model()
sm.list_model()
```

But as mentioned before, assigning `''` only to v-number is not accepted.

How to assign a constant index
How to make the pupil diameter float to the stop diameter
(I need to analysis with stop semi-diameter as variable, not entrance pupil diameter.)
How to calculate the longitudinal aberration (not transverse ray)

Sorry to ask questions in one issue. I have searched but cannot find any solutions.

Any advise is highly appreciated.

**Comment by mjhoptics** [comment]
Hello and thank you for your well explained question. I think there may be some issues with add_surface and how it's supposed to work. I'm working on addressing this case better.
The best way to approach your problem with the current code is to enter a material specification with a fake v-number, and then immediately overwrite it with a constant index material.

```
sm.gaps[0].thi=1e10
sm.add_surface([7.77, 0.55, 1.376, 56., 5.])
sm.gaps[sm.cur_surface].medium = medium.Medium(1.376, 'Medium1')
sm.add_surface([6.8, 3.4, 1.336, 56., 5.])
sm.gaps[sm.cur_surface].medium = med2 = medium.Medium(1.336, 'Medium2')
sm.add_surface([0., 1.6, 1.336, 56., 3.])
sm.gaps[sm.cur_surface].medium = med2
sm.set_stop()
sm.add_surface([20., 0.7, 1.5, 56., 3.])
sm.gaps[sm.cur_surface].medium = medium.Medium(1.5, 'Medium3')
sm.add_surface([-20., 21., 1.336, 56., 3.])
sm.gaps[sm.cur_surface].medium = med2
sm.add_surface([0., 0.])
```

It appeared from your description that there were 3 different materials used in the model; I used Medium2 in 3 places based on it's placement in your drawing.

With regards to your second question, I think the set_vignetting function might be used here. I'll think about that tomorrow.

**Reply by TLawu** [reply]
Thanks a lot Mike.
```
sm.gaps[0].thi=1e10
sm.add_surface([7.77, 0.55, 1.376, 56., 5.])
sm.gaps[sm.cur_surface].medium = rayoptics.seq.medium.Medium(1.376, 'Medium1')
sm.add_surface([6.8, 3.4, 1.336, 56., 5.])
sm.gaps[sm.cur_surface].medium = med2 = rayoptics.seq.medium.Medium(1.336, 'Medium2')
sm.add_surface([0., 1.6, 1.336, 56., 3.])
sm.gaps[sm.cur_surface].medium = med2
sm.set_stop()
sm.add_surface([20., 0.7, 1.5, 56., 3.])
sm.gaps[sm.cur_surface].medium = rayoptics.seq.medium.Medium(1.5, 'Medium3')
sm.add_surface([-20., 21., 1.336, 56., 3.])
sm.gaps[sm.cur_surface].medium = med2
sm.add_surface([0., 0.])
```
I need to use full path `rayoptics.seq.medium.Medium(1.376, 'Medium1')`
What should I do to use `medium.Medium(1.376, 'Medium1')`
However it works. Thank you.
Using fake v-number is fine with me.

I need your advice for the second and third issues (Float by Stop and Longitudinal Aberration, or Longitudinal Spherical Aberration).
There's no need to hurry. Please take your time.

I need to calculate under this condition because it is the method suggested by ISO 11979-2 for ophthalmic lens
defocus calculation, which defined the defocus as LSA/2, normally at 3 mm stop diameter.

The corrected EFL becomes f = EFL + LSA/2

In the past, I used matlab and Zemax for doing these calculations, but I want to migrate to python
and not depend on matlab and Zemax anymore.

**Comment by mjhoptics** [comment, marked as answer]
Here's a jupyter notebook addressing the 3 parts of this discussion/issue. The notebook is in the [ray-optics-notebooks repo](https://github.com/mjhoptics/ray-optics-notebooks/blob/master/Issue92_const_index.ipynb).

```python
%matplotlib inline
```


```python
from rayoptics.environment import *
from rayoptics.seq import medium    # shortcut reference to medium package
```


```python
isdark = False
```


```python
opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']   # add analysis results model for paraxial data
```


```python
osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=3)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=5.0, flds=[0.],is_relative=True)
osp['wvls'] = WvlSpec([(546.074, 1.0)], ref_wl=0)

opm.radius_mode = True
```

## 1) Model with constant index materials

### Creating a model with constant index materials

The add_surface() fct can't easily be used to create a constant refractive index material.
Use the index and a "fake" V-number to satisfy add_surface().
Then directly create and assign the constant index Medium to the appropriate gaps.


```python
sm.gaps[0].thi=1e10
sm.add_surface([7.77, 0.55, 1.376, 56., 5.])
sm.gaps[sm.cur_surface].medium = medium.Medium(1.376, 'Medium1')
sm.add_surface([6.8, 3.4, 1.336, 56., 5.])
sm.gaps[sm.cur_surface].medium = med2 = medium.Medium(1.336, 'Medium2')
sm.add_surface([0., 1.6, 1.336, 56., 3.])
sm.gaps[sm.cur_surface].medium = med2
sm.set_stop()
sm.add_surface([20., 0.7, 1.5, 56., 3.])
sm.gaps[sm.cur_surface].medium = medium.Medium(1.5, 'Medium3')
sm.add_surface([-20., 21., 1.336, 56., 3.])
sm.gaps[sm.cur_surface].medium = med2
sm.add_surface([0., 0.])
sm.ifcs[sm.cur_surface].max_aperture = 3.
sm.ifcs[-1].max_aperture = 3.
```


```python
sm.do_apertures = False
opm.update_model()
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1      1.0000
        1:     7.770000     0.550000   Medium1             1      5.0000
        2:     6.800000      3.40000   Medium2             1      5.0000
     Stop:     0.000000      1.60000   Medium2             1      3.0000
        4:    20.000000     0.700000   Medium3             1      3.0000
        5:   -20.000000      21.0000   Medium2             1      3.0000
        6:     0.000000      0.00000       air             1      3.0000
      Img:     0.000000      0.00000                       1      3.0000


## 2) Setting pupil size based on the stop size

### Use set_vignetting
The set_vignetting function ensures the bundle defining rays at each field point pass thru the edge of the most limiting aperture in the model. In this case the stop surface is the limiting aperture.


```python
set_vignetting(opm)
```

### Negative vignetting means rays overfill pupil to fill stop
Listing the object properties for the field points shows negative vignetting values. This means the entrance pupil is undersized and the vignetting is overfilling the pupil to fill the aperture stop.


```python
listobj(osp['pupil'])
listobj(osp['fov'])
```

    aperture: object pupil; value=3
    
    field: object angle; value=0.0
    x,y=0.0 vlx=-1.253 vux=-1.253 vly=-1.253 vuy=-1.253
    


### Marginal ray trace
Trace the marginal ray for the axial field point.


```python
fi = 0
fld, wvl, foc = osp.lookup_fld_wvl_focus(fi)

pupil = [0., 1.]
ray01, ray_op01, wvl = trace_base(opm, pupil, fld, wvl)
list_ray(ray01)
```

                X            Y            Z           L            M            N               Len
      0:      0.00000      0.00000            0     0.000000     0.000000     1.000000        1e+10
      1:      0.00000      3.37918      0.77329     0.000000    -0.128001     0.991774      0.63518
      2:      0.00000      3.29788      0.85324     0.000000    -0.116173     0.993229       2.5641
      3:      0.00000      3.00000            0     0.000000    -0.116173     0.993229       1.8078
      4:      0.00000      2.78998      0.19556     0.000000    -0.118727     0.992927      0.31638
      5:      0.00000      2.75242      -0.1903     0.000000    -0.150843     0.988558       21.436
      6:      0.00000     -0.48100            0     0.000000    -0.201527     0.979483            0
      7:      0.00000     -0.48100            0     0.000000    -0.201527     0.979483            0


### Revise entrance pupil diameter
Get the ray height on the first surface. For an infinite object, this is the pupil radius.


```python
rs1 = RaySeg(*ray01[1])
epd = 2*abs(rs1.p[1])
epd
```




    6.758369049845509




```python
osp['pupil'].value = epd
```


```python
opm.update_model()
```


```python
listobj(osp['fov'])
```

    field: object angle; value=0.0
    x,y=0.0 vlx=-1.253 vux=-1.253 vly=-1.253 vuy=-1.253
    


### Finish pupil redefinition
The pupil size has been redefined so the vignetting needs to be updated as well. The values to overfill the pupil are still in place. set_vignetting can fix this up.


```python
set_vignetting(opm)
```


```python
listobj(osp['pupil'])
listobj(osp['fov'])
```

    aperture: object pupil; value=6.758369049845509
    
    field: object angle; value=0.0
    x,y=0.0 vlx=-0.000 vux=-0.000 vly=-0.000 vuy=-0.000
    


### Lens layout
The lens layout doesn't look too bad. It considers the model to be a set of cemented elements for the purposes of drawing a picture.


```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,
                        do_draw_rays=True, do_draw_ray_fans=True, do_paraxial_layout=False, is_dark=isdark).plot()
```


   
    
![output_25_0](https://user-images.githubusercontent.com/33294031/180906681-30952f18-a3ed-433f-aa28-220e7c74dce2.png)



## 3) Longitudinal spherical aberration

### Calculation approach

The general approach to calculating longitudinal spherical would be to:

1) get the ray data at the image plane


```python
img_seg = RaySeg(*ray01[-1])
```


```python
img_seg.p, img_seg.d
```




    (array([ 0.        , -0.48099593,  0.        ]),
     array([ 0.        , -0.20152687,  0.97948299]))



2) divide the ray height at the image plane by the ray direction tangent.

3) I assume negative means inward from the image plane for LSA


```python
LSA = -img_seg.p[1]/(img_seg.d[1]/img_seg.d[2])
LSA
```




    -2.337789122532606




```python
efl = ar['parax_data'].fod.efl
```


```python
corrected_efl = efl + LSA/2
efl, corrected_efl
```




    (17.888918967654664, 16.72002440638836)




```python
pm.list_model()
```

               ax_ht        pr_ht       ax_slp       pr_slp         power          tau        index    type
     0:            0 -1.74551e+08  3.37918e-10    0.0174551             0        1e+10      1.00000    dummy
     1:      3.37918   -0.0590504    -0.163523    0.0203126    0.04839125     0.399709      1.37600    transmit
     2:      3.31382   -0.0509313     -0.14403     0.020013  -0.005882353      2.54491      1.33600    transmit
     3:      2.94728  1.31406e-08     -0.14403     0.020013             0       1.1976      1.33600    transmit
     4:      2.77479    0.0239677    -0.166783    0.0198165        0.0082     0.466667      1.50000    transmit
     5:      2.69696    0.0332153    -0.188898    0.0195441        0.0082      15.7186      1.33600    transmit
     6:    -0.272251      0.34042    -0.188898    0.0195441            -0            0      1.00000    transmit
     7:    -0.272251      0.34042    -0.188898    0.0195441             0            0      1.00000    dummy



```python
pm.first_order_data()
```

    efl               17.89
    ffl              -16.65
    pp1               1.242
    bfl              -1.441
    ppk               19.33
    f/#               2.647
    m            -1.789e-09
    red           -5.59e+08
    obj_dist          1e+10
    obj_ang               1
    enp_dist          3.383
    enp_radius        3.379
    na obj        3.379e-10
    n obj                 1
    img_dist         -1.441
    img_ht           0.3123
    exp_dist         -18.86
    exp_radius        3.018
    na img          -0.1856
    n img                 1
    optical invariant      0.05898


## Longitudinal Spherical Aberration plot

### Trace the rays

Use the RayFan primitive to easily trace a fan of rays across the pupil.


```python
ray_fan = analyses.RayFan(opm, f=fld, wl=wvl, xyfan='y')
```


```python
ray_list, _ = ray_fan.fan_pkg
```

### Postprocess the ray fan

Loop over the rays and calculate the LSA for each rays


```python
x = []
y = []
for r in ray_list:
    px, py, rs = r
    ray, _, _ = rs
    img = ray[-1]
    lsa = -img.p[1]/(img.d[1]/img.d[2])
    x.append(lsa)
    y.append(ray[3].p[1])  # stop surface ray height
```

### Plot using Matplotlib


```python
plt.plot(x,y)
```




    [<matplotlib.lines.Line2D at 0x7f9631e22fd0>]




    
![output_43_1](https://user-images.githubusercontent.com/33294031/180906559-a6f82107-2a80-4263-97da-ae1cd151d753.png)

**Reply by TLawu** [reply]
Hi Mike,

Great. This is exactly what I need.
Thank you very much.

**Comment by TLawu** [comment]
Dear Michael,

I hope you are doing well. It has been a while since you last provided
guidance on the constant index and setting pupil size based on the stop
size.

I want to thank you for your package, as it has been very helpful for my
work.

Recently, I upgraded my PC and installed an updated version of Python,
including newer versions of various packages. I also upgraded ray-optics to
version 0.8.7. I tried to upgrade to version 0.9.0, but encountered some
errors, so I am currently using version 0.8.7.

According to the changelog, there are new features in this version that are
relevant to my needs:

Move functionality in medium into the opticalglass, especially for working
with a constant index.
The set_pupil() function to set the pupil specification based on the
aperture stop size.
I searched the documentation, but I could not find any examples on how to
implement these features. Could you please provide me with a revised
version of the program using these new features? The program is shown below
and is also available on GitHub:
https://github.com/mjhoptics/ray-optics-notebooks/blob/master/Issue92_const_index.ipynb
.

I look forward to your reply. Thank you for your assistance.

Sincerely,
Tjundewo Lawu

**Reply by mjhoptics** [reply]
I am on the road and unable to perform any tests. Please see this [response](https://github.com/mjhoptics/ray-optics/issues/150#issuecomment-2179430523) for a problem that has been reported on numpy version.
Thanks,
Mike Hayford

**Comment by TLawu** [comment]
Dear Mike,

Take your time.
My main issue was not the version, but more with the usage of set_pupil()
and opticalglass.
I have done with opticalglass, but still don't know how to implement
set_pupil() for pupil based on STOP size.
Now, I am using your past response approach using set_vignetting()
It works, but I think it could be a more straightforward and easy approach
using set_pupil().

I hope you can assist me.
I am not in a hurry, take your time.

Thank you.
Tjundewo Lawu

**Reply by mjhoptics** [reply]
Hello @TLawu,
Thank you for your patience. Below is a notebook that shows the use of set_pupil and the resulting lens layout and longitudinal spherical aberration calculation. Please let me know if you have any questions or encounter any problems.
Mike


```python
%matplotlib inline
from rayoptics.environment import *
isdark = False
```


```python
opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']   # add analysis results model for paraxial data
```


```python
osp['pupil'] = PupilSpec(osp, key=['object', 'epd'], value=3)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=5.0, flds=[0.], is_relative=True)
osp['wvls'] = WvlSpec([(546.074, 1.0)], ref_wl=0)

opm.radius_mode = True
```

## 1) Model with constant index materials

### Creating a model with constant index materials

The add_surface() fct can' now be used to create a constant refractive index material.
In this mode, the surface semi-diameter must be entered as a keyword argument.

Note that the object and image surface are created when the OpticalModel is created, and don't need an add_surface to add them.


```python
sm.do_apertures = False
sm.gaps[0].thi=1e10
sm.add_surface([7.77, 0.55, 1.376], sd=5.)
sm.add_surface([6.8, 3.4, 1.336], sd=5.)
sm.add_surface([0., 1.6, 1.336], sd=3.)
sm.set_stop()
sm.add_surface([20., 0.7, 1.5], sd=3.)
sm.add_surface([-20., 21., 1.336], sd=3.)
```


```python
opm.update_model()
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1      1.0000
        1:     7.770000     0.550000   n:1.376             1      5.0000
        2:     6.800000      3.40000   n:1.336             1      5.0000
     Stop:     0.000000      1.60000   n:1.336             1      3.0000
        4:    20.000000     0.700000   n:1.500             1      3.0000
        5:   -20.000000      21.0000   n:1.336             1      3.0000
      Img:     0.000000      0.00000                       1      1.0000


## 2) Setting pupil size based on the stop size

There's a specific function for this now!


```python
set_pupil(opm)
```

    [0.         2.25278968]


Listing the aperture and field data indicates there pupil size has been adjusted and the field shows no vignetting.


```python
listobj(osp['pupil'])
listobj(osp['fov'])
```

    aperture: object epd; value=6.758369051519793
    
    field: object angle; value=0.0
    x,y=0.0 vlx= 0.000 vux= 0.000 vly= 0.000 vuy= 0.000
    


## Lens layout

The lens layout doesn't look too bad. It considers the model to be a set of cemented elements for the purposes of drawing a picture.


```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,
                        do_draw_rays=True, do_draw_ray_fans=True, do_paraxial_layout=False, is_dark=isdark).plot()
```


    
![output_14_0](https://github.com/user-attachments/assets/edabd0b9-14b3-4ae6-857d-c9ef0c991881)
    


## 3) Longitudinal spherical aberration

### Calculation approach

The general approach to calculating longitudinal spherical would be to:

1) get the ray data at the image plane


```python
fi = 0
fld, wvl, foc = osp.lookup_fld_wvl_focus(fi)

pupil = [0., 1.]
ray_pkg01 = trace_ray(opm, pupil, fld, wvl, output_filter=None, rayerr_filter=None)
ray01, ray_op01, wvl = ray_pkg01[0]
list_ray(ray01)
```

                X            Y            Z           L            M            N               Len
      0:      0.00000      3.37918            0     0.000000     0.000000     1.000000        1e+10
      1:      0.00000      3.37918      0.77329     0.000000    -0.128001     0.991774      0.63518
      2:      0.00000      3.29788      0.85324     0.000000    -0.116173     0.993229       2.5641
      3:      0.00000      3.00000            0     0.000000    -0.116173     0.993229       1.8078
      4:      0.00000      2.78998      0.19556     0.000000    -0.118727     0.992927      0.31638
      5:      0.00000      2.75242      -0.1903     0.000000    -0.150843     0.988558       21.436
      6:      0.00000     -0.48100            0     0.000000    -0.150843     0.988558            0



```python
img_seg = ray01[-1]
```


```python
img_seg.p, img_seg.d
```




    (array([ 0.        , -0.48099593,  0.        ]),
     array([ 0.        , -0.15084347,  0.98855766]))



2) divide the ray height at the image plane by the ray direction tangent.

3) I assume negative means inward from the image plane for LSA


```python
LSA = -img_seg.p[1]/(img_seg.d[1]/img_seg.d[2])
print(f"{LSA=:10.4f}")
```

    LSA=   -3.1522



```python
efl = ar['parax_data'].fod.efl
```


```python
corrected_efl = efl + LSA/2
print(f"{efl=:8.4f}   {corrected_efl=:8.4f}")
```

    efl= 23.8996   corrected_efl= 22.3235



```python
pm.list_model()
```

               ax_ht        pr_ht       ax_slp       pr_slp         power          tau        index    type
     0:            0 -1.74551e+08  3.37918e-10    0.0174551             0        1e+10      1.00000    dummy
     1:      3.37918   -0.0590504    -0.163523    0.0203126    0.04839125     0.399709      1.37600    transmit
     2:      3.31382   -0.0509313     -0.14403     0.020013  -0.005882353      2.54491      1.33600    transmit
     3:      2.94728  1.31406e-08     -0.14403     0.020013             0       1.1976      1.33600    transmit
     4:      2.77479    0.0239677    -0.166783    0.0198165        0.0082     0.466667      1.50000    transmit
     5:      2.69696    0.0332153    -0.188898    0.0195441        0.0082      15.7186      1.33600    transmit
     6:    -0.272251      0.34042    -0.188898    0.0195441             0            0      1.33600    dummy



```python
pm.first_order_data()
```

    efl                23.9
    f                 17.89
    f'                 23.9
    ffl              -16.65
    pp1               1.242
    bfl               19.07
    ppk              -4.825
    pp sep           0.1828
    f/#               2.647
    m            -1.789e-09
    red           -5.59e+08
    obj_dist          1e+10
    obj_ang               1
    enp_dist          3.383
    enp_radius        3.379
    na obj        3.379e-10
    n obj                 1
    img_dist          19.07
    img_ht           0.3123
    exp_dist         -4.196
    exp_radius        3.018
    na img           -0.187
    n img             1.336
    optical invariant      0.05898


## Longitudinal Spherical Aberration plot

### Trace the rays

Use the RayFan primitive to easily trace a fan of rays across the pupil.


```python
ray_fan = analyses.RayFan(opm, f=fld, wl=wvl, xyfan='y')
```


```python
ray_list, _ = ray_fan.fan_pkg
```

### Postprocess the ray fan

Loop over the rays and calculate the LSA for each rays


```python
x = []
y = []
for r in ray_list:
    px, py, rs = r
    ray, _, _ = rs
    img = ray[-1]
    lsa = -img.p[1]/(img.d[1]/img.d[2])
    x.append(lsa)
    y.append(ray[3].p[1])  # stop surface ray height
```

### Plot using Matplotlib


```python
plt.plot(x,y)
plt.show()
```


    
![output_33_0](https://github.com/user-attachments/assets/a94ed9be-7de6-4477-aa24-63f0ea51f6a8)
    



```python

```

### shift image plane to paraxial focus


```python
ax_ray = ar['parax_data'].ax_ray
```


```python
prx_img_dist = -ax_ray[-1][mc.ht]/ax_ray[-1][mc.slp]
print(f"{prx_img_dist=:10.4f}")
```

    prx_img_dist=   -1.9255



```python
sm.gaps[-1].thi += prx_img_dist
opm.update_model()
```


```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1      1.0000
        1:     7.770000     0.550000   n:1.376             1      5.0000
        2:     6.800000      3.40000   n:1.336             1      5.0000
     Stop:     0.000000      1.60000   n:1.336             1      3.0000
        4:    20.000000     0.700000   n:1.500             1      3.0000
        5:   -20.000000      19.0745   n:1.336             1      3.0000
      Img:     0.000000      0.00000                       1      1.0000



```python
ray_fan = analyses.RayFan(opm, f=fld, wl=wvl, xyfan='y')
```


```python
ray_list, _ = ray_fan.fan_pkg
```

### Postprocess the ray fan

Loop over the rays and calculate the LSA for each rays


```python
x = []
y = []
for r in ray_list:
    px, py, rs = r
    ray, _, _ = rs
    img = ray[-1]
    lsa = -img.p[1]/(img.d[1]/img.d[2])
    x.append(lsa)
    y.append(ray[3].p[1])  # stop surface ray height
```

### Plot using Matplotlib


```python
plt.plot(x,y)
plt.show()
```


    
![output_45_0](https://github.com/user-attachments/assets/6ce348e9-fe23-4c68-be83-11ee670c702d)

**Reply by TLawu** [reply]
Hi Mike,

Sorry for my late response to this issue, since I need to travel abroad.
I tried your code above but got errors, and don't know how to solve. Please advice.

Currently, I use the following environments:
python Python 3.12.5
rayoptics 0.9.1
numpy 1.26.4

In process 1) Model with constant index materials
I got erroer on opm.update_model().
File ~\miniconda3\lib\site-packages\rayoptics\parax\firstorder.py:255, in compute_first_order(opt_model, stop, wvl)
    253         slpk = n_k*math.tan(math.asin(pupil.value/n_k))
    254         slp0 = slpk/red
--> 255 yu = [0., slp0]
    257 yu_bar = [1., 0.]
    258 fov = opt_model.optical_spec.field_of_view

UnboundLocalError: local variable 'slp0' referenced before assignment

Regards,
Tjundewo

**Reply by mjhoptics** [reply]
Hello @TLawu,
I'm sorry you're having this difficulty. I have not been able to reproduce the problem with the sample code I have here. I do think I've seen this error but I can't recall enough of the circumstances to reproduce it. If you could provide you're steps, that would be helpful.
Thanks,
Mike Hayford

**Reply by TLawu** [reply]
Hello Mike,

Sorry for the trouble.
I checked again, and found the problem.
I didn't install the environment for raypotics version 0.9.1 in the Jupyter Notebook.
After correcting the environment, everything run correctly.

Thank you and sorry again for my carelessness.
Best regards,
Tjundewo

**Comment by TLawu** [comment]
Dear Mike,

Thank you for your kindness to help me out.
I will work with your advice and will reach you again if I am facing any
problems.
I think it will be good for now.

Best regards,
Tjundewo Lawu

**Reply by TLawu** [reply]
Dear Mike,

I have a question.
When pupil size is based on STOP size, then set_pupil(opm) is used.
What to do to set back to pupil size based on Entrance Pupil Diameter?
I mean switching from STOP to EPD and vice versa.

Regards,
Tjundewo

**Reply by mjhoptics** [reply]
set_pupil(opm) is a function that determines the aperture on the stop surface. Internally, it sets osp['pupil'].value to the EPD determined by the stop diameter. You can use the same approach to directly set the EPD to another value.
```
osp['pupil'].value = 6.0
opm.update_model()
```

**Reply by TLawu** [reply]
Dear Mike,

Thank you.

Tjundewo

**Reply by TLawu** [reply]
Dear Mike,

Sorry to trouble you again.
I found error with set_pupil(opm) when it involved Air.
There are 2 issues.
The first one is print command in the set_pupil() function. I can suppress this output with contexmanager.
The second one is the following error, when the sm have surface with Air.
Error only when using set_pupil(opm) but no error when setting osp['pupil'].value

Please help me solve this problem.
Thank you and best regards,
Tjundewo


```python
%matplotlib inline
import sys
import os
import contextlib   # for not print the set_pupil(opm)
from rayoptics.environment import *
isdark = False

@contextlib.contextmanager
def suppress_stdout():
    null_stream = open(os.devnull, 'w')
    old_stdout = sys.stdout
    try:
        sys.stdout = null_stream
        yield
    finally:
        sys.stdout = old_stdout
        null_stream.close()
``` 

```python
opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']   # add analysis results model for paraxial data
``` 

```python
osp['pupil'] = PupilSpec(osp, key=['object', 'epd'], value=3)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=5.0, flds=[0.], is_relative=True)
osp['wvls'] = WvlSpec([(546.074, 1.0)], ref_wl=0)
opm.radius_mode = True
``` 

```python
sm.do_apertures = False
sm.gaps[0].thi=1e10
sm.add_surface([19.332, 10., 1.493], sd=8)
sm.add_surface([0., 3.])  # This line will give error with set_pupil(opm)
# sm.add_surface([0., 3., 1.3], sd=8)  # If this line is used then no error, but this is not correct medium. It should be Air.
sm.add_surface([0, 6., 1.519], sd=8)
sm.add_surface([0, 6.5, 1.336], sd=8)
sm.set_stop()
sm.add_surface([0, 10., 1.336], sd=8)
sm.add_surface([0, 6., 1.519], sd=8)
sm.add_surface([0., 3.1712])
``` 

```python
opm.update_model()
``` 

```python
# Setting pupil size based on the stop size
with suppress_stdout():
    set_pupil(opm)
opm.update_model()
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,
                        do_draw_rays=True, do_draw_ray_fans=True, do_paraxial_layout=False, is_dark=isdark).plot()
``` 

I got the following error
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[6], line 3
      1 # Setting pupil size based on the stop size
      2 with suppress_stdout():
----> 3     set_pupil(opm)
      4 opm.update_model()
      5 layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm,
      6                         do_draw_rays=True, do_draw_ray_fans=True, do_paraxial_layout=False, is_dark=isdark).plot()

File ~\miniconda3\envs\rayoptics_env\Lib\site-packages\rayoptics\gui\appcmds.py:390, in set_pupil(opt_model, gui_parent)
    388 def set_pupil(opt_model, gui_parent=None):
    389     """ From existing stop size, calculate pupil spec and vignetting. """
--> 390     vigcalc.set_pupil(opt_model)
    391     if gui_parent is None:
    392         opt_model.update_model(src_model=opt_model['seq_model'])

File ~\miniconda3\envs\rayoptics_env\Lib\site-packages\rayoptics\raytr\vigcalc.py:103, in set_pupil(opm)
    101 if obj_img_key == 'object':
    102     if pupil_spec == 'epd' or pupil_spec == 'pupil':
--> 103         rs1 = RaySeg(*ray_pkg[0][1])
    104         ht = rs1.p[1]
    105         osp['pupil'].value = 2*ht

TypeError: 'NoneType' object is not subscriptable
```

**Reply by TLawu** [reply]
Dear Mike,

The problem was solved.
I made mistake in setting sd value for STOP. If I changed
`sm.add_surface([0, 6.5, 1.336], sd=8)` to `sm.add_surface([0, 6.5, 1.336], sd=1.5)` before `sm.set_stop()`
then no errors.

Sorry for the trouble.

Best regards,
Tjundewo

**Reply by TLawu** [reply]
Additionally, sd must be set in `sm.add_surface([0., 3.], sd=8.)`

**Reply by mjhoptics** [reply]
Tjundewo - glad to hear you got it debugged. Mike
